WhateverGreen/kern_igfx_memory.cpp

```diff
@@ -0,0 +1,209 @@
+	// As such, this patch locates instructions related to calculating the amount of memory available and identifies which register stores the final result.
+	// It calculates the correct number of bytes beforehand and finds the proper instruction to copy the value to the result register.
+	// Consequently, the framebuffer controller will initialize the memory manager with proper values and aforementioned kernel panics can be avoided.
+	//
+	// - FireWolf
+	// - 2020.08
+	//
+	hde64s handle;
+	uint64_t shllAddr = 0, andlAddr = 0; /* Instruction Address  */
+	uint32_t shllSize = 0, andlSize = 0; /* Instruction Length   */
+	uint32_t shllDstr = 0, andlDstr = 0; /* Destination Register */
+	
+	// e.g. movl $0x03C00000, %eax (60MB DVMT)
+	// Apply the middle 5 bytes if the target register is %eax;
+	// Apply the last 6 bytes if the target register is below %r8d;
+	// Apply all 7 bytes if the target register is or above %r8d.
+	uint8_t movl[] = {0x41, 0xB8, 0x00, 0x00, 0x00, 0x00, 0x90};
+	uint8_t nops[] = {0x90, 0x90, 0x90, 0x90};
+	
+	// Patch Heuristics:
+	// We need to locate two instructions `shll` and `andl` that manipulate the value read from the GGC field,
+	// but we cannot assume that they always stay together and operate on the register %eax.
+	// As such, we need to first figure out which register is of interest, and
+	// we need 5 or 6 bytes to move a 32-bit integer to the register manipulated by the above instructions.
+	// Since `andl` is 5 - 7 bytes long, we could just replace it with a "movl" and then erases `shll` by filling `nop`s.
+	for (auto index = 0; index < 64; index += 1) {
+		auto size = Disassembler::hdeDisasm(startAddress, &handle);
+		
+		// Guard: Should be able to disassemble the function
+		if (handle.flags & F_ERROR) {
+			SYSLOG("igfx", "DVMT: Failed to disassemble FBMemMgr_Init().");
+			break;
+		}
+		
+		// Instruction: shll $0x11, %???
+		// 3 bytes long if DSTReg < %r8d, otherwise 4 bytes long
+		if (handle.opcode == 0xC1 && handle.imm.imm8 == 0x11) {
+			shllAddr = startAddress;
+			shllSize = handle.len;
+			shllDstr = (handle.rex_b << 3) | handle.modrm_rm;
+			SYSLOG("igfx", "DVMT: Found the shll instruction. Length = %d; DSTReg = %d.", shllSize, shllDstr);
+		}
+		
+		// Instruction: andl $0xFE000000, %???
+		// 5 bytes long if DSTReg is %eax; 6 bytes long if DSTReg < %r8d; otherwise 7 bytes long.
+		if ((handle.opcode == 0x25 || handle.opcode == 0x81) && handle.imm.imm32 == 0xFE000000) {
+			andlAddr = startAddress;
+			andlSize = handle.len;
+			andlDstr = (handle.rex_b << 3) | handle.modrm_rm;
+			SYSLOG("igfx", "DVMT: Found the andl instruction. Length = %d; DSTReg = %d.", andlSize, andlDstr);
+		}
+		
+		// Guard: Calculate and apply the binary patch if we have found both instructions
+		if (shllAddr && andlAddr) {
+			// Update the `movl` instruction with the actual amount of DVMT preallocated memory
+			*reinterpret_cast<uint32_t*>(movl + 2) = dvmt;
```

Comment:
Is this preceded with `setKernelWriting`?

Reply:
See above: https://github.com/acidanthera/WhateverGreen/pull/64#discussion_r480917666.